File: medium/2126-destroying-asteroids/solution.py
from typing import List


class Solution:
    def asteroidsDestroyed(self, mass: int, asteroids: List[int]) -> bool:
        """
        2126. Destroying Asteroids
        Time: O(n + M)
        Space: O(M)
        """
        max_asteroid = max(asteroids)
        frequency = [0] * (max_asteroid + 1)

        for asteroid in asteroids:
            frequency[asteroid] += 1

        for asteroid_mass, count in enumerate(frequency):
            if count == 0:
                continue

            if mass < asteroid_mass:
                return False
            mass += asteroid_mass * count

        return True

File: medium/2126-destroying-asteroids/solution.java
/**
 * 2126. Destroying Asteroids
 * Time: O(n + M)
 * Space: O(M)
 */
class Solution {
    public boolean asteroidsDestroyed(int mass, int[] asteroids) {
        int maxAsteroid = 0;
        for (int asteroid : asteroids) {
            maxAsteroid = Math.max(maxAsteroid, asteroid);
        }

        int[] frequency = new int[maxAsteroid + 1];
        for (int asteroid : asteroids) {
            frequency[asteroid]++;
        }

        long currentMass = mass;

        for (int asteroidMass = 1; asteroidMass <= maxAsteroid; asteroidMass++) {
            if (frequency[asteroidMass] == 0) {
                continue;
            }

            if (currentMass < asteroidMass) {
                return false;
            }
            currentMass += (long) asteroidMass * frequency[asteroidMass];
        }

        return true;
    }
}
